### PR TITLE
[ts] Add support for `expr satisfies Type` expressions

### PR DESCRIPTION
**Pull request by nicolo-ribaudo**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | Fixes: #15067
| Patch: Bug Fix?          |
| Major: Breaking Change?  |
| Minor: New Feature?      |
| Tests Added + Pass?      | Yes
| Documentation PR Link    | 
| Any Dependency Changes?  |
| License                  | MIT

https://github.com/microsoft/TypeScript/pull/46827 is in the [TypeScript 4.6 iteration plan](https://github.com/microsoft/TypeScript/issues/46858): I'm opening this PR so that we are ready in case it gets merged before Babel 7.17.0. However, I'm marking this PR as a draft because that one is still marked as a draft (@DanielRosenwasser, do you think that there is any chance that that PR will be actually included in 4.6, even if you already released a beta without it?).

---

I'm not convinced by the new AST node I introduced for `expr satisfies Type` expressions. It feels like it should be the same node as `TSAsExpression`, similarly to how `in` and `instanceof` both are `BinaryExpression`s. Ideally, something like this:
```ts
type TSBinaryExpression = { // or TSTypeExpression
  type: "TSBinaryExpression",
  operator: "as" | "satisfies",
  expression: Expression,
  typeAnnotation: TSType,
}
```
however, that would be a breaking change. I also considered just adding `operator` to `TSAsExpression`, but `{ type: "TSAsExpression", operator: "satisfies" }` looks wrong.

Maybe we could go for the breaking change in Babel 8, and for now provide a `TSBinaryExpression` alias in `@babel/types`?

`TSBinaryExpression` or `TSTypeExpression`? The new node would _not_ cover `<Type> expr`, because it needs to visit its childrens in a different order (type before expression).

<a href="https://gitpod.io/#https://github.com/babel/babel/pull/14211"></a>

### CODE REVIEW

File: packages/babel-parser/src/plugins/typescript/index.js [hunk 1]
```diff
@@ -2278,23 +2274,35 @@ export default (superClass: Class<Parser>): Class<Parser> =>
...
-          node.typeAnnotation = _const;
-        } else {
-          node.typeAnnotation = this.tsNextThenParseType();
-        }
-        this.finishNode(node, "TSAsExpression");
+        node.typeAnnotation = this.tsInType(() => {
+          this.next(); // "as" or "satisfies"
+          if (this.match(tt._const)) {
+            if (isSatisfies) {
+              this.raise(
```
Comment (JLHwung) [inline packages/babel-parser/src/plugins/typescript/index.js:2291]: I think we don't have to specialize `tt._const` after `tt._satisfies`: `tsParseType` should throw when `const` is seen.

**Comment by DanielRosenwasser** [comment]
Just as a heads up, this never made its way into TypeScript 4.6 - it's still being iterated on.

**Comment by nicolo-ribaudo** [comment]
Thank you!

**Comment by nicolo-ribaudo** [comment]
The TS pull request has been merged, I'll rebase this.

**Comment by nicolo-ribaudo** [comment]
Depending on the 7.19 timing, we might need to postpone this to 7.20.

File: packages/babel-generator/src/generators/typescript.ts [hunk 2]
```diff
@@ -522,15 +522,23 @@ export function TSTypeAliasDeclaration(
...
-  const { expression, typeAnnotation } = node;
+function TSTypeExpression(
+  this: Printer,
+  node: t.TSAsExpression | t.TSSatisfiesExpression,
+) {
+  const { type, expression, typeAnnotation } = node;
   this.print(expression, node);
   this.space();
-  this.word("as");
+  this.word(type === "TSAsExpression" ? "as" : "satisfies");
```
Comment (JLHwung) [inline packages/babel-generator/src/generators/typescript.ts:532]: We should ensure no line breaking is allowed between `expression` and `as | satisfies`.

File: packages/babel-plugin-transform-typescript/src/index.ts [hunk 3]
```diff
@@ -593,11 +593,14 @@ export default declare((api, opts: Options) => {
         path.replaceWith(path.node.expression);
       },
 
-      TSAsExpression(path) {
+      [`TSAsExpression${
+        // Added in Babel 7.17.0
```
Comment (JLHwung) [inline packages/babel-plugin-transform-typescript/src/index.ts:597]:
```suggestion
        // Added in Babel 7.20.0
```

**Review by JLHwung — APPROVED** [review]

File: packages/babel-parser/src/plugins/typescript/index.ts [hunk 4]
```diff
@@ -721,26 +721,6 @@ export default (superClass: ClassWithMixin<typeof Parser, IJSXParserMixin>) =>
...
-
-      this.next();
-      const typeReference = this.tsParseTypeReference();
-
-      // If the type reference has type parameters, then you are using it as a
-      // type and not as a const signifier. We'll *never* be able to find this
-      // name, since const isn't allowed as a type name. So in this instance we
-      // get to pretend we're the type checker.
-      if (typeReference.typeParameters) {
-        this.raise(TSErrors.CannotFindName, {
```
Comment (liuxingbaoyu) [inline packages/babel-parser/src/plugins/typescript/index.ts:735]: `CannotFindName` seems to be used only here.

File: packages/babel-types/src/builders/generated/index.ts [hunk 5]
```diff
@@ -2351,6 +2352,17 @@ export function tsAsExpression(
...
+  expression: t.Expression,
+  typeAnnotation: t.TSType,
+): t.TSSatisfiesExpression {
+  return validateNode<t.TSSatisfiesExpression>({
+    type: "TSSatisfiesExpression",
+    expression,
+    typeAnnotation,
+  });
+}
+export { tsSatisfiesExpression as tSSatisfiesExpression };
```
Comment (liuxingbaoyu) [inline packages/babel-types/src/builders/generated/index.ts:2365]: This seems to be a bug.😕

**Review by liuxingbaoyu — APPROVED** [review]

**Comment by thorn0** [comment]
"Type expression" sounds like something that operates on types only whereas in `as` and `satisfies` one operand is a value and the other one a type.

**Comment by ngregory-rbi** [comment]
> "Type expression" sounds like something that operates on types only...

I disagree. An adjective does not describe an operand but a property of the thing itself. In math, we don't call the addition symbol a "number expression". It operates on numbers, it itself is not a number.


This is a "type expression" because it expresses a type to either:
- cast the operand value to, or
- ensure the operand value satisfies

Makes sense to me.

**Comment by nicolo-ribaudo** [comment]
The exposed names are TSAsExpression and TSSatisfiesExpression; TSTypeExpression is just an internal variable name.

**Comment by thorn0** [comment]
@ngregory-rbi You have a point, but in TypeScript there is this dichotomy between values and types. And when it's needed to specify one of these contexts, this kind of wording is used. E.g. "type operators" [here](https://www.typescriptlang.org/docs/handbook/2/types-from-types.html).

On the other hand, I just noticed that the TS team strictly avoids using the word "expression" for type-only things, which makes `TSTypeExpression` a better choice than I thought.

> we don't call the addition symbol a "number expression". It operates on numbers, it itself is not a number.

Can't help myself, excuse me :) , but what about "dog food" then? Is it a dog?

**Comment by ngregory-rbi** [comment]
> You have a point, but in TypeScript there is this dichotomy between values and types. And when it's needed to specify one of these contexts, this kind of wording is used. E.g. "type operators" [here](https://www.typescriptlang.org/docs/handbook/2/types-from-types.html).

Fair enough.

> On the other hand, I just noticed that the TS team strictly avoids using the word "expression" for type-only things, which makes TSTypeExpression a better choice than I thought.

Not sure how your conclusion connects to the preceding statement.

> Can't help myself, excuse me :) , but what about "dog food" then? Is it a dog?

Fair point as well. Touché.

Regardless of anything, wording may be a tangential subject to this PR.

**Comment by thorn0** [comment]
> Not sure how your conclusion connects to the preceding statement.

The preceding statement was against applying the wording "type expression" to `as` and `satisfies` whereas the following one was in support. Both statements are a result of checking what wording is used in the TS docs.

@nicolo-ribaudo There are your questions about naming in the first post here, so I thought they were still relevant.

**Comment by nicolo-ribaudo** [comment]
Oh thanks, I forgot about it :upside_down_face: 

Re-reading my original post, I don't like TSBinaryExpression because it feels like it should be an expression between two TS things, but this one is between a JS value and a type. For Babel 8, TSTypeExpression might be fine.

**Comment by ngregory-rbi** [comment]
@thorn0 I meant more the two segments here:

> On the other hand, I just noticed that the TS team strictly avoids using the word "expression" for type-only things, which makes TSTypeExpression a better choice than I thought.

If the TS team "strictly avoids" using the word "expression", why does it make Babel using the term a "better choice than [you] thought"? Would you not want to keep the same wording as the TS team?

**Review by JLHwung — APPROVED** [review]

**Comment by ngregory-rbi** [comment]
Please merge this as soon as possible :)

**Comment by nicolo-ribaudo** [comment]
Release tomorrow!

**Pull request merged**